src/osmet.rs

```diff
@@ -0,0 +1,969 @@
+    }
+
+    let repo = Path::new(&config.repo);
+    println!("Writing image to device");
+    let n = write_unpunched_image(&mut punched_image, &mut device, &osmet.partitions, repo)?;
+    device.flush()?;
+
+    if n != osmet.size {
+        bail!("wrote {} bytes but expected {}", n, osmet.size);
+    }
+
+    println!("Verifying digest");
+    let checksum = get_path_digest_limited(Path::new(&config.device), Some(osmet.size))
+        .chain_err(|| format!("while hashing {}", &config.device))?;
+
+    if checksum != osmet.checksum {
+        bail!("final device checksum does not match");
+    }
+
+    device.flush()?;
+    reread_partition_table(&mut device)?;
+    println!("Unpacking successful!");
+
+    Ok(())
+}
+
+fn get_unpunched_image_digest(
+    punched_image: &mut File,
+    partitions: &[OsmetPartition],
+    root: &Mount,
+) -> Result<(Sha256Digest, u64)> {
+    let mut hasher = Sha256::new();
+    let repo = {
+        let mut root_path = root.mountpoint().to_path_buf();
+        root_path.push("ostree/repo");
+        root_path
+    };
+
+    let mut unxz_punched_image = XzDecoder::new(punched_image);
+    let n = write_unpunched_image(&mut unxz_punched_image, &mut hasher, &partitions, &repo)?;
+    unxz_punched_image.into_inner().seek(SeekFrom::Start(0))?;
+    Ok((Sha256Digest(hasher.result().into()), n))
+}
+
+fn write_unpunched_image(
+    punched_image: &mut impl Read,
+    w: &mut impl Write,
+    partitions: &[OsmetPartition],
+    repo: &Path,
+) -> Result<u64> {
+    let mut buf = [0u8; 8192];
+
+    // start streaming writes to device, interspersing OSTree objects
+    let mut cursor: u64 = 0;
+    for partition in partitions {
+        assert!(partition.start_offset >= cursor);
+        cursor += copy_exactly_n(punched_image, w, partition.start_offset - cursor, &mut buf)?;
+        cursor += write_partition(w, partition, punched_image, repo, &mut buf)?;
+    }
+
+    // and copy the rest
+    cursor += io::copy(punched_image, w)?;
+
+    Ok(cursor)
+}
+
+fn scan_root_partition(
+    root: &Mount,
+    mut boot_files: HashMap<u64, PathBuf>,
+) -> Result<(OsmetPartition, HashMap<PathBuf, Sha256Digest>)> {
+    // query the trivial stuff first
+    let (start_offset, end_offset) = root.get_partition_offsets()?;
+
+    // we only hash boot files if there's a potential match with an OSTree object, so we keep a
+    // cache to avoid recomputing it multiple times
+    let mut cached_boot_files_digests: HashMap<u64, Sha256Digest> = HashMap::new();
+
+    // boot files we were able to match up with OSTree objects
+    let mut mapped_boot_files: HashMap<PathBuf, Sha256Digest> = HashMap::new();
+
+    let objects_dir = {
+        let mut root_path = root.mountpoint().to_path_buf();
+        root_path.push("ostree/repo/objects");
+        root_path
+    };
+
+    let mut mappings: Vec<Mapping> = vec![];
+    let mut mapped_file_count = 0;
+    let mut empty_file_count = 0;
+    for entry in WalkDir::new(objects_dir) {
```

Comment:
I'm fine with this but... I think it's also poking into ostree internals.  It's not like we are likely to change the layout (doing so would be *really hard* due to backcompat constraints) but...consider using  https://crates.io/crates/ostree maybe?

Reply:
I considered this a bit, but `ostree_repo_list_objects` is basically doing the same thing except it's incurring way more overhead from handling all object types and allocating lots of `GVariant`s.

The other thing is that, we don't really care about what commits are in the repo for example. Whatever file object we find is clearly on the disk and so would benefit from mapping. That said, querying libostree for commit information could still be used before unpacking as a sanity-check that we'll have all the objects we need. I can definitely see linking to libostree in this code in the future though. Just not sure if we need it just yet.